회원 프로필 이미지 수정 / 삭제 기능

seongheonsim opened the pull request:
@holeman79 
안녕하세요, 원진님! 테스트 코드 정리를 하느라고 조금 늦었습니다ㅠㅠ
구현한 기능은 다음과 같아요.

- 회원 프로필 수정 API : `POST /api/member/{id}/profile`
- 회원 프로필 삭제 API : `DELETE /api/member/{id}/profile`
- 업로드 파일 조회 API : `GET /api/download/{fileName}`

파일 경로는 추후에 프로필 이미지 뿐만 아니라 빙고 아이템에도 이미지 등록될 수 있는 경우를 고려해 `FILE_INFO` 라는 테이블에서 관리하도록 구현 했습니다. 때문에 `MEMBERS` 테이블과 `FILE_INFOS` 는 연관관계를 맺고 있으며, 프로필 변경 및 삭제 시 기존 연결된 레코드는 삭제하도록 했습니다.

현재 파일은 공통 도메인이라고 생각해서 `/global/domain` 디렉토리에 위치하도록 하고 여러 계층을 모두 `/global/domain/file` 디렉토리 내에 구현 했는데, 이게 과연 좋은 방법인지 확신이 서질 않네요..🥲 해당 부분에 대해서 원진님 의견이 어떠신지 궁금합니다..!

그리고 추가적인 변경사항은 다음과 같습니다.

- `CustomDescriptor.kt` 
  - PathParameters 에 대한 infix 함수명과 구조를 수정했습니다. 이로 인해 해당 infix 함수를 호출하는 코드가 변경되었습니다.
  - RequestParts 에 대한 infix 함수가 추가되었습니다.
- Snippets
  -  PathParameters 스니팻을 오버라이트하는 파일을 추가했습니다. (`path-parameters.snippet`)
  -  RequestParts 스니팻을 오버라이트하는 파일을 추가했습니다. (`request-parts.snippet`)

Commit:
feat(MemberProfile) : 회원 프로필 이미지 등록/삭제 기능 추가

- File 도메인 관련 공통 클래스 추가 (/global/domain/file)

Commit:
feat(FIleInfo) : 테이블명 수정

holeman79 commented on src/main/kotlin/com/beside/groubing/groubingserver/domain/member/api/MemberProfileEditApi.kt:
```diff
@@ -0,0 +1,27 @@
+        @PathVariable id: Long,
+        @RequestPart profile: MultipartFile
+    ): ApiResponse<MemberProfileResponse> {
+        val profileUrl = memberProfileEditService.deleteAndEdit(id, profile)
```
deleteAndEdit 이란 이름보다 그냥 edit 또는 등록이라는 명칭이 낫지 않을까요? 기존 프로필 이미지를 지우는 작업 때문에 넣으신 것 같은데 어쩔 수 없는 경우가 아니면 메소드 명칭은 and 없이 사용하는게 좋습니다.
예를 들어 은행 계좌이체 메소드를 만든다고 했을 때 
minusMyAccountAndPlusOtherAccount 이런식으로 작성하지 않고 transferAccount 이런 명칭으로 사용하듯이요ㅎㅎ

seongheonsim replied:
네, 저도 그 부분 고민하다가 `deleteAndEdit` 로 결정 했는데 원진님 예시를 생각해보니 `edit` 으로만 하는 것이 가독성에 더 좋을 것 같습니다! 해당 부분 반영 완료했습니다 :')

holeman79 commented on src/main/kotlin/com/beside/groubing/groubingserver/domain/member/domain/Member.kt:
```diff
@@ -32,6 +37,10 @@ class Member internal constructor(
     var nickname: String = nickname
         private set
 
+    @ManyToOne(fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
```
profile 이미지는 OneToOne 아닌가요? 똑같은 이미지 파일이 여러 사용자에게 사용될 일이 있나요?

seongheonsim replied:
FileInfoRepository 를 통해 삭제 메소드를 호출하지 않으려고 하다보니 연관관계를 ManyToOne 으로 설정했습니다.
하지만 원진님 말씀하신대로 똑같은 이미지 파일이 여러 사용자에게 사용될 일이 없으니 오히려 ManyToOne 으로 관계를 맺는게 옳지 않는게 맞는 것 같아요..!😂

해당 부분 OneToOne 관계로 수정하고, 회원 프로필 이미지 삭제 서비스에서 FileInfoRepository 를 주입 받아 delete 메소드 호출하도록 수정했습니다~

holeman79 commented on src/main/kotlin/com/beside/groubing/groubingserver/global/domain/file/api/FileDownloadApi.kt:
```diff
@@ -0,0 +1,26 @@
+import org.springframework.web.bind.annotation.RestController
+
+@RestController
+@RequestMapping("/api/download")
```
file 관련 기능도 마찬가지로 rest api 규칙대로 작성하는게 어떨까요? 
GET api/files/{filename} -> 파일 다운로드
POST api/files -> 파일 업로드
이런 형태로요!

seongheonsim replied:
@holeman79 
그러면 유저 프로필 업로드 API 는 제거하고 `/api/files` 로 했을 때 유저 프로필 설정은 어떻게 하면 좋을까요?
다운로드 API 는 원진님 말씀하신대로 반영하는게 좋을 것 같은데, 만약 유저 프로필 이미지 뿐만 아니라 추후 빙고 아이템에 이미지 등록도 고려했을 때 어떻게 확장해야 할지 고민이 됩니다ㅠㅠ

holeman79 replied:
유저 프로필 이미지는 파일 업로드 뿐만 아니라 유저의 profile 속성값도 변해야하기 때문에 post api/files 로 호출처리 하지 않고 PUT or PATCH api/members/profile 이런 형태로 처리해야하지 않을까 싶습니다. 사실 딱 file만 업로드 하는 경우는 잘 없어서 POST api files 호출하는 경우는 거의 없긴 하겠네요. GET api/files/{filename} 만 있으면 될 것 같습니다ㅎㅎ

seongheonsim replied:
@holeman79 
원진님, 해당 내용에 대한 피드백이 늦어져 죄송합니다.
먼저 파일 다운로드에 대한 API 규칙은 수정해서 반영 완료 했습니다.
추가적으로 `RestDocumentationRequestBuilders` 클래스에서 호출하는 `multipart()` 메소드는 오직 `POST` 만 테스트가 가능하도록 되어 있는데, 이를 `CustomMockMvcRequestBuilders` 파일에 확장함수를 추가하여 `PATCH` 로도 테스트 가능하게끔 추가 했습니다. 한번 확인 부탁드립니다!

Commit:
feat(FileDownloadApi) : REST API 규칙 반영

Commit:
feat(MemberProfileEditService) : 유저 프로필 이미지 수정 함수명 변경

Commit:
feat(FileDownloadApi) : API Endpoint 수정

Commit:
feat(Member) : Member <-> FileInfo 연관관계 수정

Commit:
feat(Member) : 프로필 수정 Http Method 수정 및 테스트 코드 변경

- CustomMockMvcRequestBuilders 클래스 추가하여 multipart() 메소드에서 PUT 및 PATCH 도 테스트 진행할 수 있도록 반영
